Source/Particles/ParticleBoundaryBuffer.cpp

```diff
@@ -0,0 +1,150 @@
+            pp_species.query("scrape_zhi", m_do_boundary_buffer[5][ispecies]);
+#endif
+        }
+    }
+}
+
+void ParticleBoundaryBuffer::printNumParticles () const {
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim)
+    {
+        for (int iside = 0; iside < 2; ++iside)
+        {
+            auto& buffer = m_particle_containers[2*idim+iside];
+            const int nspecies = nSpecies();
+            for (int i = 0; i < nspecies; ++i)
+            {
+                int np;
+                if (buffer[i].isDefined()) { np = buffer[i].TotalNumberOfParticles(false); }
+                else { np = 0; }
+                amrex::Print() << "Species " << m_species_names[i] << " has "
+                               << np << " particles in the boundary buffer "
+                               << " for side " << iside << " of dim " << idim << "\n";
+            }
+        }
+    }
+}
+
+struct IsOutsideBoundary {
+    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> m_plo;
+    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> m_phi;
+    int m_idim;
+    int m_iside;
+
+    template <typename SrcData>
+    AMREX_GPU_DEVICE
+    int operator() (const SrcData& src,
+                    int ip, const amrex::RandomEngine& /*engine*/) const noexcept
+    {
+        const auto& p = src.getSuperParticle(ip);
+        if (m_iside == 0) {
+            if (p.pos(m_idim) < m_plo[m_idim]) { return 1; }
+        } else {
+            if (p.pos(m_idim) >= m_phi[m_idim]) { return 1; }
+        }
+        return 0;
+    }
+};
+
+void ParticleBoundaryBuffer::gatherParticles (MultiParticleContainer& mypc) {
+    const auto& warpx_instance = WarpX::GetInstance();
+    const amrex::Geometry& geom = warpx_instance.Geom(0);
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim)
+    {
+        if (geom.isPeriodic(idim)) continue;
+        for (int iside = 0; iside < 2; ++iside)
+        {
+            auto plo = geom.ProbLoArray();
+            auto phi = geom.ProbHiArray();
+            auto& buffer = m_particle_containers[2*idim+iside];
+            const int nspecies = mypc.nSpecies();
+            for (int i = 0; i < nspecies; ++i)
+            {
+                if (!m_do_boundary_buffer[2*idim+iside][i]) continue;
+                const auto& pc = mypc.GetParticleContainer(i);
+                if (!buffer[i].isDefined())
+                {
+                    buffer[i] = ParticleBuffer::getTmpPC<amrex::PinnedArenaAllocator>(&pc);
+                }
+                auto& species_buffer = buffer[i];
+                species_buffer.addParticles(pc, IsOutsideBoundary{plo, phi, idim, iside}, true);
+            }
+        }
+    }
+}
+
+int ParticleBoundaryBuffer::getNumParticlesInContainer(std::string species_name, int boundary) {
+    if (m_do_boundary_buffer[boundary].size() == 0) {
+        return 0;
+    }
+
+    auto& buffer = m_particle_containers[boundary];
+
+    std::vector<std::string> all_species = m_species_names;
```

Comment:
As with the return type comment, this should be a reference rather than a copy. I'd suggest just using the member directly, rather than assigning a local variable, though.